addon/content/reducer.js

```diff
@@ -107,7 +128,167 @@ const attachmentActions = {
+      "message.movedOrDeletedConversation"
+    );
+  } else {
+    window.Conversations = {
+      currentConversation: null,
+      counter: 0,
+    };
+    let freshConversation = new Conversation(
+      window,
+      // TODO: This should really become ids at some stage, but we need to
+      // teach Conversation how to handle those.
+      msgUrls,
+      scrollMode,
+      ++Conversations.counter,
+      isInTab
+    );
+    let browserFrame = window.frameElement;
+    // Because Thunderbird still hasn't fixed that...
+    if (browserFrame) {
+      browserFrame.setAttribute("context", "mailContext");
+    }
+
+    freshConversation.outputInto(window, async function(aConversation) {
+      // This is a stripped-down version of what's in monkeypatch.js,
+      //  make sure the two are in sync!
+      Conversations.currentConversation = aConversation;
+      aConversation.completed = true;
+      // TODO: Re-enable this.
+      // registerQuickReply();
+      // That's why we saved it before...
+      // newComposeSessionByDraftIf();
+      // TODO: expandQuickReply isn't defined anywhere. Should it be?
+      // if (willExpand)
+      //   expandQuickReply();
+      // Create a new rule that will override the default rule, so that
+      // the expanded quick reply is twice higher.
+      document.body.classList.add("inTab");
+      // Do this now so as to not defeat the whole expand/collapse
+      // logic.
+      if (
+        await browser.conversations.getCorePref(
+          "mailnews.mark_message_read.auto"
+        )
+      ) {
+        setTimeout(function() {
+          for (const id of msgIds) {
+            browser.messages.update(id, { read: true }).catch(console.error);
+          }
+        }, (await browser.conversations.getCorePref(
+          "mailnews.mark_message_read.delay.interval"
+        )) *
+          (await browser.conversations.getCorePref(
+            "mailnews.mark_message_read.delay"
+          )) *
+          1000);
+      }
+    });
+  }
+}
+
 const messageActions = {
+  waitForStartup() {
+    return async dispatch => {
+      const params = new URL(document.location).searchParams;
+
+      const isInTab = params.has("urls");
+
+      // These can happen at the same time.
+      await Promise.all([
+        browser.runtime.getPlatformInfo().then(info => {
+          dispatch({
+            type: "SET_OS",
+            OS: info.os,
+          });
+        }),
+        dispatch({
+          type: "SET_IN_TAB",
+          isInTab,
+        }),
+      ]);
+
+      if (!isInTab) {
+        return;
+      }
+
+      await new Promise(resolve => {
+        function checkStarted() {
+          let mainWindow = topMail3Pane(window);
+          if (
+            mainWindow.Conversations &&
+            mainWindow.Conversations.monkeyPatch &&
+            mainWindow.Conversations.monkeyPatch.finishedStartup
+          ) {
+            resolve();
+          } else {
+            setTimeout(checkStarted, 100);
```

Comment:
Do we want to add a loop counter here, or trigger a warning if things are taking a really long time?

Reply:
Yeah we shouldn't hit it, but I've added a trap just in case. I went for 10 seconds as startup could be a case where we get delayed a bit.